src/ModelContextProtocol/Protocol/Types/Resources/ResourceContents.cs

```diff
@@ -0,0 +1,120 @@
+﻿using System.Text.Json;
+using System.Text.Json.Serialization;
+
+namespace ModelContextProtocol.Protocol.Types;
+
+/// <summary>
+/// The contents of a specific resource or sub-resource.
+/// <see href="https://github.com/modelcontextprotocol/specification/blob/main/schema/2024-11-05/schema.json">See the schema for details</see>
+/// </summary>
+[JsonConverter(typeof(ResourceContentsConverter))]
+public abstract class ResourceContents
+{
+    /// <summary>
+    /// The URI of the resource.
+    /// </summary>
+    [JsonPropertyName("uri")]
+    public string Uri { get; set; } = string.Empty;
+
+    /// <summary>
+    /// The type of content.
+    /// </summary>
+    [JsonPropertyName("mimeType")]
+    public string? MimeType { get; set; }
+}
+
+internal class ResourceContentsConverter : JsonConverter<ResourceContents>
```

Comment:
It needs to be public to work with the source generator. I think in the past we've made them EditorBrowsableState.Never, and possibly nested them within the type they're converting. @eiriktsarpalis ?

Reply:
What will the source generator be needed for in touching this?

Reply:
> What will the source generator be needed for in touching this?

The source generated by the source generator needs to be able to access the converter. If the serialization is being done in this assembly, internal is fine. But if another assembly using the source generator wants to serialize this type, the converter needs to be accessible to that generated source.